File: src/chunk.cpp
#include "chunk.hpp"
#include "chunk_system.hpp"
#include "command.hpp"
#include "preview_system.hpp"
#include "room.hpp"
#include "server.hpp"
#include "session.hpp"
#include <cassert>

Chunk::Chunk(ChunkSystem *chunk_system, Int2 position, SharedVector<u8> compressed_chunk_data)
		: chunk_system(chunk_system),
			position(position) {
	LockGuard lock(mtx_access);
	this->compressed_image = compressed_chunk_data;

	new_chunk = true;
	if(compressed_chunk_data && !compressed_chunk_data->empty())
		new_chunk = false;
}

Chunk::~Chunk() {
	while(!linked_sessions.empty()) {
		fprintf(stderr, "Chunk linked sessions NOT empty");
		abort();
	}
}

u32 Chunk::getImageSizeBytes() const {
	return ChunkSystem::getChunkSize() * ChunkSystem::getChunkSize() * 3; /*RGB*/
}

void Chunk::allocateImage_nolock() {
	if(!image) {
		image = createSharedVector<u8>(getImageSizeBytes());
		new_chunk = false;

		if(compressed_image) {
			decompressLZ4(compressed_image->data(), compressed_image->size(), image->data(), image->size());
		} else {
			memset(image->data(), 255, image->size()); // White
		}
	}
}

void Chunk::getPixel_nolock(UInt2 chunk_pixel_pos, Color *color) {
	assert(chunk_pixel_pos.x < ChunkSystem::getChunkSize());
	assert(chunk_pixel_pos.y < ChunkSystem::getChunkSize());
	auto *rgb = image->data();
	size_t offset = chunk_pixel_pos.y * ChunkSystem::getChunkSize() * 3 + chunk_pixel_pos.x * 3;
	color->r = rgb[offset + 0];
	color->g = rgb[offset + 1];
	color->b = rgb[offset + 2];
}

static SharedVector<u8> compressed_empty_chunk;
static Mutex mtx_empty_chunk;

// Returns LZ4-compressed empty, white chunk. Generates once.
SharedVector<u8> getEmptyChunk(Chunk *chunk) {
	LockGuard lock(mtx_empty_chunk);
	if(!compressed_empty_chunk) {
		uniqdata<u8> stub_img(chunk->getImageSizeBytes());
		memset(stub_img.data(), 255, stub_img.size_bytes());
		compressed_empty_chunk = compressLZ4(stub_img.data(), stub_img.size_bytes());
	}
	return compressed_empty_chunk;
}

SharedVector<u8> Chunk::encodeChunkData_nolock() {
	SharedVector<u8> compressed;

	if(new_chunk) {
		// Return compressed empty chunk
		compressed = getEmptyChunk(this);
	} else {
		allocateImage_nolock();
		compressed = compressLZ4(image->data(), image->size());
	}

	this->compressed_image = compressed;
	return compressed;
}

SharedVector<u8> Chunk::encodeChunkData(bool clear_modified) {
	LockGuard lock(mtx_access);
	auto compressed = encodeChunkData_nolock();
	if(clear_modified) {
		setModified_nolock(false);
		image.reset();

		auto *preview_system = chunk_system->room->getPreviewSystem();

		Int2 upper_pos;
		upper_pos.x = position.x >= 0 ? position.x / 2 : (position.x - 1) / 2;
		upper_pos.y = position.y >= 0 ? position.y / 2 : (position.y - 1) / 2;
		preview_system->addToQueueFront(upper_pos);
	}
	return compressed;
}

void Chunk::sendChunkDataToSession_nolock(Session *session) {
	SharedVector<u8> compressed_data;

	if(compressed_image) {
		// Grab from cache
		compressed_data = compressed_image;
	} else {
		// Recompress chunk data
		compressed_data = encodeChunkData_nolock();
	}

	s32 chunk_x_BE = tobig32((s32)getPosition().x);
	s32 chunk_y_BE = tobig32((s32)getPosition().y);
	u32 raw_size_BE = tobig32((u32)getImageSizeBytes());

	Datasize data_chunk_x(&chunk_x_BE, sizeof(s32));
	Datasize data_chunk_y(&chunk_y_BE, sizeof(s32));
	Datasize data_raw_size(&raw_size_BE, sizeof(u32));
	Datasize data_compressed_data(compressed_data->data(), compressed_data->size());

	Datasize *datasizes[] = {
			&data_chunk_x,
			&data_chunk_y,
			&data_raw_size,
			&data_compressed_data,
			nullptr};

	session->pushPacket(preparePacket(ServerCmd::chunk_image, datasizes));
}

void Chunk::linkSession(Session *session) {
	LockGuard lock(mtx_access);

	// Check if session pointer already exists
	for(auto &cell : linked_sessions) {
		if(cell == session)
			return;
	}

	linked_sessions_empty = false;

	// Add session pointer
	linked_sessions.push_back(session);

	sendChunkDataToSession_nolock(session);
}

void Chunk::unlinkSession(Session *session) {
	LockGuard lock(mtx_access);

	// Find pointer
	for(auto it = linked_sessions.begin(); it != linked_sessions.end();) {
		if(*it == session) {
			// Remove session pointer
			it = linked_sessions.erase(it);
			break;
		} else {
			it++;
		}
	}

	bool is_empty = linked_sessions.empty();
	linked_sessions_empty = is_empty;

	if(is_empty)
		chunk_system->markGarbageCollect();
}

bool Chunk::isLinkedSessionsEmpty() {
	return linked_sessions_empty;
}

void Chunk::lock() {
	mtx_access.lock();
}

void Chunk::unlock() {
	mtx_access.unlock();
}

void Chunk::setPixelsQueued_nolock(ChunkPixel *pixels, u32 count) {
	allocateImage_nolock();
	auto *rgb = image->data();

	if(!send_chunk_data_instead_of_pixels) {
		queued_pixels_to_send.reserve(queued_pixels_to_send.size() + count);
	}

	for(u32 i = 0; i < count; i++) {
		auto &pixel = pixels[i];
		size_t offset = pixel.pos.y * ChunkSystem::getChunkSize() * 3 + pixel.pos.x * 3;
		rgb[offset + 0] = pixel.color.r;
		rgb[offset + 1] = pixel.color.g;
		rgb[offset + 2] = pixel.color.b;

		if(!send_chunk_data_instead_of_pixels) {
			queued_pixels_to_send.push_back(pixel);
			if(queued_pixels_to_send.size() > 5000) {
				queued_pixels_to_send = {};
				send_chunk_data_instead_of_pixels = true;
			}
		}
	}
	setModified_nolock(true);
}

void Chunk::setPixelQueued_nolock(ChunkPixel *pixel) {
	setPixelsQueued_nolock(pixel, 1);
}

void Chunk::setPixelQueued(ChunkPixel *pixel) {
	LockGuard lock(mtx_access);
	setPixelQueued_nolock(pixel);
}

void Chunk::flushQueuedPixels() {
	LockGuard lock(mtx_access);
	flushQueuedPixels_nolock();
}

void Chunk::flushQueuedPixels_nolock() {
	if(send_chunk_data_instead_of_pixels) {
		for(auto &session : linked_sessions) {
			sendChunkDataToSession_nolock(session);
		}
		send_chunk_data_instead_of_pixels = false;
	} else {
		if(queued_pixels_to_send.empty()) return;
		setPixels_nolock(queued_pixels_to_send.data(), queued_pixels_to_send.size(), true);
		queued_pixels_to_send.clear();
	}
}

void Chunk::setPixels(ChunkPixel *pixels, size_t count) {
	LockGuard lock(mtx_access);
	flushQueuedPixels_nolock();
	setPixels_nolock(pixels, count);
}

void Chunk::setPixels_nolock(ChunkPixel *pixels, size_t count, bool only_send) {
	allocateImage_nolock();

	// Prepare pixel_pack packet
	Buffer buf_pixels;
	u32 pixel_count = 0;

	Color color;
	for(size_t i = 0; i < count; i++) {
		auto &pixel = pixels[i];

		if(!only_send) {
			getPixel_nolock(pixel.pos, &color);
			if(pixel.color == color) {
				// Pixel not changed, skip
				continue;
			}

			// Update pixel
			auto *rgb = image->data();
			size_t offset = pixel.pos.y * ChunkSystem::getChunkSize() * 3 + pixel.pos.x * 3;
			rgb[offset + 0] = pixel.color.r;
			rgb[offset + 1] = pixel.color.g;
			rgb[offset + 2] = pixel.color.b;
		}

		// Prepare pixel data
		u8 x = pixel.pos.x;
		u8 y = pixel.pos.y;

		buf_pixels.write(&x, sizeof(u8));
		buf_pixels.write(&y, sizeof(u8));
		buf_pixels.write(&pixel.color.r, sizeof(u8));
		buf_pixels.write(&pixel.color.g, sizeof(u8));
		buf_pixels.write(&pixel.color.b, sizeof(u8));
		pixel_count++;
	}

	if(pixel_count == 0)
		return; // Nothing modified

	auto compressed = compressLZ4(buf_pixels.data(), buf_pixels.size());

	s32 chunk_x_BE = tobig32((s32)getPosition().x);
	s32 chunk_y_BE = tobig32((s32)getPosition().y);
	u32 pixel_count_BE = tobig32((u32)pixel_count);
	u32 raw_size_BE = tobig32((u32)buf_pixels.size());

	Datasize data_chunk_x(&chunk_x_BE, sizeof(s32));
	Datasize data_chunk_y(&chunk_y_BE, sizeof(s32));
	Datasize data_pixel_count(&pixel_count_BE, sizeof(u32));
	Datasize data_raw_size(&raw_size_BE, sizeof(u32));
	Datasize data_compressed_data(compressed->data(), compressed->size());

	Datasize *datasizes[] = {
			&data_chunk_x,
			&data_chunk_y,
			&data_pixel_count,
			&data_raw_size,
			&data_compressed_data,
			nullptr};

	auto packet = preparePacket(ServerCmd::chunk_pixel_pack, datasizes);

	for(auto &session : linked_sessions) {
		session->pushPacket(packet);
	}

	setModified_nolock(true);
}

Int2 Chunk::getPosition() const {
	return position;
}

bool Chunk::isModified() {
	return modified;
}

void Chunk::setModified_nolock(bool n) {
	modified = n;
	if(modified) {
		// Compressed image data is now invalid
		compressed_image.reset();
	}
}

File: src/chunk.rs
use std::sync::{Arc, OnceLock, Weak};

use bytes::{BufMut, BytesMut};
use glam::{IVec2, UVec2};
use std::sync::Mutex as SyncMutex;
use tokio::sync::Mutex;

use crate::{
	compression::{self, compress_lz4},
	event_queue::EventQueue,
	gen_id,
	limits::{self, CHUNK_SIZE_PX},
	packet_server::{self, prepare_packet_pixel_pack},
	pixel::Color,
	preview_system::PreviewSystemQueuedChunks,
	session::{SessionHandle, SessionInstanceWeak},
	signal::Signal,
};

#[derive(Clone)]
pub struct ChunkPixel {
	pub pos: UVec2,
	pub color: Color,
}

pub struct LinkedSession {
	_session: SessionInstanceWeak,
	queue_send: EventQueue<packet_server::Packet>,
	handle: SessionHandle,
}

#[derive(Clone)]
pub struct ChunkInstanceRefs {
	pub modified: Arc<SyncMutex<bool>>,
	pub linked_sessions: Arc<SyncMutex<Vec<LinkedSession>>>,
}

pub struct ChunkInstance {
	new_chunk: bool,
	pub position: IVec2,
	refs: ChunkInstanceRefs,
	pub raw_image_data: Option<Vec<u8>>,
	compressed_image_data: Option<Arc<Vec<u8>>>,
	preview_system_queued_chunks: PreviewSystemQueuedChunks,
	signal_garbage_collect: Signal,
}

// Returns LZ4-compressed empty, white chunk. Generates once.
fn get_empty_chunk() -> &'static std::sync::Mutex<Arc<Vec<u8>>> {
	static CHUNK_DATA: OnceLock<std::sync::Mutex<Arc<Vec<u8>>>> = OnceLock::new();
	CHUNK_DATA.get_or_init(|| {
		let mut stub_img: Vec<u8> = Vec::new();
		stub_img.resize(limits::CHUNK_IMAGE_SIZE_BYTES, 255);
		std::sync::Mutex::new(Arc::new(compression::compress_lz4(&stub_img)))
	})
}

impl ChunkInstance {
	pub fn new(
		position: IVec2,
		refs: ChunkInstanceRefs,
		preview_system_queued_chunks: PreviewSystemQueuedChunks,
		signal_garbage_collect: Signal,
		compressed_image_data: Option<Vec<u8>>,
	) -> Self {
		Self {
			new_chunk: compressed_image_data.is_some(),
			refs,
			position,
			preview_system_queued_chunks,
			signal_garbage_collect,
			compressed_image_data: compressed_image_data.map(Arc::new),
			raw_image_data: None,
		}
	}

	pub fn allocate_image(&mut self) {
		if self.raw_image_data.is_some() {
			return; // Nothing to do, already allocated
		}

		self.new_chunk = false;

		if let Some(compressed) = &self.compressed_image_data {
			// Decode compressed data
			if let Some(raw) = compression::decompress_lz4(compressed, limits::CHUNK_IMAGE_SIZE_BYTES) {
				self.raw_image_data = Some(raw);
				return;
			}
		}

		// Failed to load image, allocate white chunk
		let mut data: Vec<u8> = Vec::new();
		data.resize(limits::CHUNK_IMAGE_SIZE_BYTES, 255); // White color
		self.raw_image_data = Some(data);
	}

	fn set_modified(&mut self, modified: bool) {
		*self.refs.modified.lock().unwrap() = modified;
		if modified {
			// Invalidate compressed data
			self.compressed_image_data = None;
		}
	}

	pub fn encode_chunk_data(&mut self, clear_modified: bool) -> Arc<Vec<u8>> {
		let compressed = if self.new_chunk {
			// Return compressed empty chunk
			return get_empty_chunk().lock().unwrap().clone();
		} else {
			self.allocate_image();
			let raw_image = self.raw_image_data.as_ref().unwrap();
			let data = Arc::new(compression::compress_lz4(raw_image));
			self.compressed_image_data = Some(data.clone());
			data
		};

		if clear_modified {
			self.set_modified(false);
			self.raw_image_data = None;

			let upper_pos = IVec2::new(
				if self.position.x >= 0 {
					self.position.x / 2
				} else {
					(self.position.x - 1) / 2
				},
				if self.position.y >= 0 {
					self.position.y / 2
				} else {
					(self.position.y - 1) / 2
				},
			);
			self.preview_system_queued_chunks.send(upper_pos);
		}

		compressed
	}

	/// Use allocate_image() first before calling this function.
	pub fn get_pixel(&self, chunk_pixel_pos: UVec2) -> Color {
		debug_assert!(self.raw_image_data.is_some());

		let offset = (chunk_pixel_pos.y * CHUNK_SIZE_PX * 3 + chunk_pixel_pos.x * 3) as usize;

		if let Some(data) = &self.raw_image_data {
			return Color {
				r: data[offset],
				g: data[offset + 1],
				b: data[offset + 2],
			};
		}

		unreachable!();
	}

	pub fn set_pixels(&mut self, pixels: &[ChunkPixel], only_send: bool, send_whole_chunk: bool) {
		debug_assert!(self.raw_image_data.is_some());

		if send_whole_chunk {
			for pixel in pixels {
				// Update pixel
				let offset = (pixel.pos.y * CHUNK_SIZE_PX * 3 + pixel.pos.x * 3) as usize;
				if let Some(data) = &mut self.raw_image_data {
					data[offset] = pixel.color.r;
					data[offset + 1] = pixel.color.g;
					data[offset + 2] = pixel.color.b;
				} else {
					unreachable!()
				}
			}

			self.set_modified(true);

			let session_send_queues: Vec<_> = self
				.refs
				.linked_sessions
				.lock()
				.unwrap()
				.iter()
				.map(|session| session.queue_send.clone())
				.collect();
			for send_queue in session_send_queues {
				self.send_chunk_data_to_session(send_queue);
			}
		} else {
			// Prepare pixel_pack packet
			let mut buf = BytesMut::new();
			let mut pixel_count = 0;

			for pixel in pixels {
				if !only_send {
					let color = self.get_pixel(pixel.pos);
					if pixel.color == color {
						// Pixel not changed, skip
						continue;
					}

					// Update pixel
					let offset = (pixel.pos.y * CHUNK_SIZE_PX * 3 + pixel.pos.x * 3) as usize;
					if let Some(data) = &mut self.raw_image_data {
						data[offset] = pixel.color.r;
						data[offset + 1] = pixel.color.g;
						data[offset + 2] = pixel.color.b;
					} else {
						unreachable!()
					}
				}

				// Prepare pixel data
				buf.put_u8(pixel.pos.x as u8);
				buf.put_u8(pixel.pos.y as u8);
				buf.put_u8(pixel.color.r);
				buf.put_u8(pixel.color.g);
				buf.put_u8(pixel.color.b);
				pixel_count += 1;
			}

			if pixel_count == 0 {
				// Nothing modified
				return;
			}

			self.set_modified(true);

			// LZ4-compressed (xyrgb,xyrgb,xyrgb...) data
			let compressed_buf = compress_lz4(&buf);

			let packet = prepare_packet_pixel_pack(
				&compressed_buf,
				self.position.x,
				self.position.y,
				pixel_count,
				buf.len() as u32,
			);

			for session in self.refs.linked_sessions.lock().unwrap().iter() {
				session.queue_send.send(packet.clone());
			}
		}
	}

	pub fn send_chunk_data_to_session(
		&mut self,
		session_queue_send: EventQueue<packet_server::Packet>,
	) {
		let compressed_data = if let Some(compressed) = &self.compressed_image_data {
			compressed.clone()
		} else {
			self.encode_chunk_data(false)
		};

		session_queue_send.send(packet_server::prepare_packet_chunk_image(
			self.position,
			&compressed_data,
		));
	}

	pub fn link_session(
		&mut self,
		handle: &SessionHandle,
		session: SessionInstanceWeak,
		session_queue_send: EventQueue<packet_server::Packet>,
	) {
		let mut linked_sessions = self.refs.linked_sessions.lock().unwrap();

		for s in linked_sessions.iter() {
			if s.handle == *handle {
				log::error!("Session is already linked!");
			}
		}

		linked_sessions.push(LinkedSession {
			handle: *handle,
			_session: session.clone(),
			queue_send: session_queue_send,
		});
	}

	pub fn unlink_session(&mut self, handle: &SessionHandle) {
		let mut to_remove_idx: Option<usize> = None;

		let mut linked_sessions = self.refs.linked_sessions.lock().unwrap();

		for (idx, session) in linked_sessions.iter().enumerate() {
			if session.handle == *handle {
				to_remove_idx = Some(idx);
				break;
			}
		}

		if let Some(idx) = to_remove_idx {
			linked_sessions.remove(idx);
		} else {
			log::warn!("Cannot unlink non-existent session");
		}

		if linked_sessions.is_empty() {
			self.signal_garbage_collect.notify();
		}
	}
}

pub type ChunkInstanceMutex = Arc<Mutex<ChunkInstance>>;
pub type ChunkInstanceWeak = Weak<Mutex<ChunkInstance>>;
gen_id!(ChunkVec, ChunkInstanceMutex, ChunkCell, ChunkHandle);
